unchained-commons/src/main/java/unchained/commons/contract/Factory.java

```diff
@@ -0,0 +1,22 @@
+package unchained.commons.contract;
+
+import unchained.commons.utils.ReflectionUtils;
+
+import java.util.function.Function;
+
+@FunctionalInterface
+public interface Factory<T> extends Function<Object[], T> {
+
+    T create(Object... arguments);
+
+    @Override
+    default T apply(Object[] arguments) {
+        return create(arguments);
+    }
+
+    @SuppressWarnings("unchecked")
+    default Class<T> type() {
```

Comment:
`Class<T>` does not capture all type situations, specifically generics: `Factory<List<String>>` will be resolved to `Class<List<String>>` which is statically equal to `Class<List<?>>`.

Using `Type` might help.